### PR TITLE
Add Initial Mixins

### PR DESCRIPTION
**Pull request by itdependsnetworks**

### CODE REVIEW

File: docs/mixins.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+* Creates the `_unique_data` attribute to be able to stuff data into
+* Super class's add method, to automatically update `_unique_data` as needed.
+
+## AdapterMixIn
+
+The `AdapterMixIn` provides two features:
+
+* apply_model_flags - Provides a convenience method to apply model flags.
+* *apply_diffsync_flags - Provides a convenience method to apply model flags based on tags on the nautobot model.
```
Comment (chadell) [inline docs/mixins.md:44]:
```suggestion
* apply_diffsync_flags - Provides a convenience method to apply model flags based on tags on the nautobot model.
```

File: docs/mixins.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    * e.g. `_many_to_many = {"tagged_vlans": "vlan"}` where `tagged_vlans` is an attribute on the current model.
+* `_generic_relation` is in the format of the key is the current model attribute and value of a dictionary. The dictionary needs to include 3 keys, `parent` to identify the model of the parent, `identifiers` is a list of the identifiers, and `attr` is the attribute to be set to.
+    * e.g. `_generic_relation = {"interface": {"parent": "interface", "identifiers": ["device", "interface"], "attr": "assigned_object"}}`
+
+
+## DiffSyncMixIn
+
+The `DiffSyncMixIn` provides two features:
+
+* Creates the `_unique_data` attribute to be able to stuff data into
```
Comment (chadell) [inline docs/mixins.md:36]:
```suggestion
* Creates the `_unique_data` attribute to be able to stuff data into.
```

File: nautobot_ssot/mixins.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+    @classmethod
+    def create(cls, diffsync, ids, attrs):
+        """Create Method to handle generically."""
+        model = super().create(ids=ids, diffsync=diffsync, attrs=attrs)
+        # model is the DiffSyncModel instance, which has data about the instance required
+        db_model = cls.Meta.model
+        obj = db_model()
+        instance_values = {**ids, **attrs}
+        combined = (
```
Comment (chadell) [inline nautobot_ssot/mixins.py:38]: `combined` what? maybe better `combined_refernce_keys`? Trying to get a name that better describes the intention and helps to understand

File: docs/mixins.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+As the class name indicates, this is intended to be used to extend a `DiffSyncModel` based class. There are 5 private attributes (private to avoid namespace and Pydantic overwrite) that are used to control how the CRUD operations works. The defintion of the attributes is:
+
+```python
+class DiffSyncModelMixIn:
+
+    _foreign_key = {}
+    _many_to_many = {}
+    _generic_relation = {}
+    _unique_fields = ("pk",)
+    _skip = ()
```
Comment (chadell) [inline docs/mixins.md:16]: in the code is `_skip = []`

File: nautobot_ssot/mixins.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            if key in list(self._foreign_key.keys()):
+                db_obj = get_fk(self._foreign_key, diffsync, key, val)
+                _vars[key] = str(db_obj.pk)
+            if key in list(self._many_to_many.keys()):
+                db_obj = get_fk(self._many_to_many, diffsync, key, val)
+                _vars[key] = str(db_obj.pk)
+
+        obj = db_model.objects.get(**_vars)
+
+        for key, value in instance_values.items():
```
Comment (chadell) [inline nautobot_ssot/mixins.py:92]: this code looks really similar to the one from `create`

File: nautobot_ssot/mixins.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+                    else:
+                        db_obj = get_fk(self._many_to_many, diffsync, key, value)
+                        obj.set(db_obj)
+        obj.save()
+        return super().update(attrs)
+
+    def delete(self):
+        """Create Method to handle generically."""
+        db_model = self.Meta.model
+        obj = db_model.objects.get(pk=self.pk)
```
Comment (chadell) [inline nautobot_ssot/mixins.py:126]: are we 100% sure it exists?

File: nautobot_ssot/mixins.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    """Mixin to update add to allow for 'stuffing' data based on unique fields."""
+
+    _unique_data = {}
+
+    def add(self, obj, *args, **kwargs):
+        """Override add method to stuff data into dictionary based on the `_unique_fields`."""
+        super().add(obj, *args, **kwargs)
+        modelname = obj._modelname
+
+        for attr in getattr(obj, "_unique_fields", []):
```
Comment (chadell) [inline nautobot_ssot/mixins.py:143]: shouldn't be `unique_data`? (or the other way around)

File: nautobot_ssot/mixins.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+            if hasattr(obj, attr):
+                if not self._unique_data.get(modelname):
+                    self._unique_data[modelname] = {}
+                if not self._unique_data[modelname].get(attr):
+                    self._unique_data[modelname][attr] = {}
+                self._unique_data[modelname][attr][getattr(obj, attr)] = obj
+
+
+class AdapterMixIn:
+    def apply_diffsync_flags(self):
```
Comment (chadell) [inline nautobot_ssot/mixins.py:153]: isn't this something in diffsync? it seem to me like something for diffsync library

**Comment by itdependsnetworks** [comment]
@chadell going to resolve your comments for now, a lot of things will change so best to start once out of draft.

**Comment by Kircheneer** [comment]
Superseded by #153

**Pull request closed without merging**